Firestore/core/src/firebase/firestore/util/CMakeLists.txt

```diff
@@ -72,41 +62,49 @@ cc_select(
 )
 
 
-## log
+## base
 
 cc_library(
-  firebase_firestore_util_log_stdio
+  firebase_firestore_util_base_stdio
   SOURCES
     hard_assert_stdio.cc
     hard_assert.h
     log.h
     log_stdio.cc
+    string_format.cc
+    string_format.h
+    string_win.cc
+    string_win.h
   DEPENDS
     firebase_firestore_util_base
     absl_base
+    absl_strings
   EXCLUDE_FROM_ALL
 )
 
 cc_library(
-  firebase_firestore_util_log_apple
+  firebase_firestore_util_base_apple
   SOURCES
     hard_assert.h
     hard_assert_apple.mm
     log.h
     log_apple.mm
     string_apple.h
+    string_format.cc
+    string_format.h
   DEPENDS
     FirebaseCore
     GoogleUtilities
+    absl_base
     absl_strings
     firebase_firestore_util_base
   EXCLUDE_FROM_ALL
 )
 
 cc_select(
-  firebase_firestore_util_log
-  APPLE   firebase_firestore_util_log_apple
-  DEFAULT firebase_firestore_util_log_stdio
+  firebase_firestore_util_base
```

Comment:
Since you're already switch'ing on the os, another alternative would be to introduce firebase_firestore_util_base_win32. This would keep _stdio free of platform specific stuff. (Yeah, the reasons here are different. I'm not sure if this is actually useful.)

Optional.

Reply:
In general if _apple and _win code *supplements* the platform independent code it's OK to include it alongside that code. We need to fork the build graph when the platform-specific code replaces the platform independent code.

log_apple and log_stdio require a link-time switch, but string_apple and string_win can coexist. I'd prefer to keep this as simple as possible and only fork the build graph for implementations that are ODR-incompatible.

Reply:
sgtm